File: src/lib/scrapers/workday.ts
/**
 * Workday ATS Scraper
 *
 * Fetches job listings from Workday career sites.
 * API: POST https://{tenant}.{server}.myworkdayjobs.com/wday/cxs/{tenant}/{site}/jobs
 */

import {
  stripHtml,
  extractSalary,
  retryWithBackoff,
  generateSourceId,
  delay,
} from './utils';
import type { ScrapedJobData } from './greenhouse';

// ==================== Types ====================

export interface WorkdayJob {
  title: string;
  externalPath: string;
  locationsText: string;
  postedOn: string;
  bulletFields: string[];
}

export interface WorkdayJobsResponse {
  total: number;
  jobPostings: WorkdayJob[];
}

export interface WorkdayJobDetail {
  jobPostingInfo: {
    id: string;
    title: string;
    jobDescription: string;
    location?: string;
    timeType?: string;
    postedOn?: string;
    jobRequisitionId?: string;
    externalUrl?: string;
  };
}

export interface WorkdayConfig {
  tenant: string;
  server: string;  // e.g., "wd12", "wd1", "wd5"
  site: string;    // e.g., "External_Career_Site"
}

// ==================== Constants ====================

const RATE_LIMIT_DELAY_MS = 200; // 0.2s between requests - Workday is fast
const PAGE_SIZE = 20; // Workday default

// ==================== Scraper Functions ====================

/**
 * Parse Workday slug format: "tenant.server.site"
 * Example: "salesforce.wd12.External_Career_Site"
 */
function parseWorkdaySlug(slug: string): WorkdayConfig {
  const parts = slug.split('.');
  if (parts.length !== 3) {
    throw new Error(`Invalid Workday slug format: ${slug}. Expected: tenant.server.site`);
  }
  return {
    tenant: parts[0],
    server: parts[1],
    site: parts[2],
  };
}

/**
 * Build Workday API base URL
 */
function buildWorkdayUrl(config: WorkdayConfig): string {
  return `https://${config.tenant}.${config.server}.myworkdayjobs.com/wday/cxs/${config.tenant}/${config.site}`;
}

/**
 * Fetch all jobs from a Workday career site
 */
export async function fetchWorkdayJobs(
  slug: string,
  companyName: string
): Promise<ScrapedJobData[]> {
  const config = parseWorkdaySlug(slug);
  const baseUrl = buildWorkdayUrl(config);
  const jobs: ScrapedJobData[] = [];
  let offset = 0;
  let total = 0;

  do {
    const url = `${baseUrl}/jobs`;

    const response = await retryWithBackoff(
      async () => {
        const res = await fetch(url, {
          method: 'POST',
          headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json',
            'User-Agent': 'GenzJobs/1.0 (job aggregator)',
          },
          body: JSON.stringify({
            appliedFacets: {},
            limit: PAGE_SIZE,
            offset: offset,
          }),
        });

        if (!res.ok) {
          if (res.status === 404) {
            throw new Error(`Workday site not found: ${slug}`);
          }
          if (res.status === 429) {
            throw new Error('Rate limited by Workday API');
          }
          throw new Error(`Workday API error: ${res.status} ${res.statusText}`);
        }

        return res.json() as Promise<WorkdayJobsResponse>;
      },
      {
        maxRetries: 3,
        initialDelayMs: 1000,
        maxDelayMs: 30000,
      }
    );

    total = response.total;

    // Process jobs in this batch
    for (const job of response.jobPostings) {
      // Fetch full job details
      const detail = await fetchJobDetail(baseUrl, job.externalPath);

      if (detail) {
        const scraped = parseWorkdayJob(job, detail, slug, companyName, config);
        if (scraped) {
          jobs.push(scraped);
        }
      }

      // Small delay between job detail fetches
      await delay(RATE_LIMIT_DELAY_MS);
    }

    offset += PAGE_SIZE;
  } while (offset < total);

  return jobs;
}

/**
 * Fetch job detail from Workday
 */
async function fetchJobDetail(
  baseUrl: string,
  externalPath: string
): Promise<WorkdayJobDetail | null> {
  try {
    const url = `${baseUrl}${externalPath}`;
    const res = await fetch(url, {
      headers: {
        'Accept': 'application/json',
        'User-Agent': 'GenzJobs/1.0 (job aggregator)',
      },
    });

    if (!res.ok) return null;

    return await res.json() as WorkdayJobDetail;
  } catch {
    return null;
  }
}

/**
 * Parse a Workday job into our standard format
 */
function parseWorkdayJob(
  listing: WorkdayJob,
  detail: WorkdayJobDetail,
  slug: string,
  companyName: string,
  config: WorkdayConfig
): ScrapedJobData | null {
  const info = detail.jobPostingInfo;

  // Strip HTML from description
  const description = stripHtml(info.jobDescription || '');

  // Skip jobs without description
  if (!description || description.length < 50) {
    return null;
  }

  // Parse location
  const location = listing.locationsText || info.location || null;

  // Detect remote
  const remote = location?.toLowerCase().includes('remote') ?? false;

  // Extract salary from description
  const salaryInfo = extractSalary(description);

  // Parse job type from timeType
  const jobType = parseTimeType(info.timeType);

  // Build apply URL
  const applyUrl = `https://${config.tenant}.${config.server}.myworkdayjobs.com/${config.site}${listing.externalPath}`;

  // Parse posted date
  let postedAt = new Date();
  if (listing.postedOn) {
    if (listing.postedOn.includes('Today')) {
      postedAt = new Date();
    } else if (listing.postedOn.includes('Yesterday')) {
      postedAt = new Date(Date.now() - 24 * 60 * 60 * 1000);
    } else if (listing.postedOn.includes('30+ Days Ago')) {
      postedAt = new Date(Date.now() - 30 * 24 * 60 * 60 * 1000);
    } else {
      // Try to parse "Posted X Days Ago"
      const daysMatch = listing.postedOn.match(/(\d+)\s*Days?\s*Ago/i);
      if (daysMatch) {
        postedAt = new Date(Date.now() - parseInt(daysMatch[1]) * 24 * 60 * 60 * 1000);
      }
    }
  }

  // Extract job ID from bulletFields or externalPath
  const jobId = listing.bulletFields?.[0] ||
    listing.externalPath.split('_').pop() ||
    info.id;

  return {
    sourceId: generateSourceId('workday', config.tenant, jobId),
    source: 'workday',
    title: listing.title,
    company: companyName,
    location,
    remote,
    description,
    applyUrl,
    salaryMin: salaryInfo?.min || null,
    salaryMax: salaryInfo?.max || null,
    salaryCurrency: salaryInfo?.currency || null,
    salaryPeriod: salaryInfo?.period || null,
    jobType,
    department: null, // Would need to parse from description or add facet filtering
    postedAt,
  };
}

/**
 * Parse Workday timeType into standardized job type
 */
function parseTimeType(timeType: string | undefined): string | null {
  if (!timeType) return null;

  const lower = timeType.toLowerCase();

  if (lower.includes('full')) return 'full-time';
  if (lower.includes('part')) return 'part-time';
  if (lower.includes('contract') || lower.includes('temp')) return 'contract';
  if (lower.includes('intern')) return 'internship';

  return null;
}

/**
 * Validate a Workday site exists
 */
export async function validateWorkdaySite(slug: string): Promise<boolean> {
  try {
    const config = parseWorkdaySlug(slug);
    const baseUrl = buildWorkdayUrl(config);

    const response = await fetch(`${baseUrl}/jobs`, {
      method: 'POST',
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json',
        'User-Agent': 'GenzJobs/1.0 (job aggregator)',
      },
      body: JSON.stringify({
        appliedFacets: {},
        limit: 1,
        offset: 0,
      }),
    });

    return response.ok;
  } catch {
    return false;
  }
}

/**
 * Get rate limit delay for Workday
 */
export function getWorkdayRateLimitDelay(): number {
  return 500; // 0.5s between companies
}

File: scraper/scrapers/workday.py
import httpx
import logging
from typing import List, Optional
from datetime import datetime

from .ats_base import ATSBaseScraper
from models import ScrapedJob
from companies import WORKDAY_COMPANIES
from config import MAX_JOBS_PER_SOURCE

logger = logging.getLogger(__name__)

PAGE_SIZE = 20  # Workday default and max per request


class WorkdayScraper(ATSBaseScraper):
    """Scraper for Workday ATS boards.
    API: POST https://{tenant}.{server}.myworkdayjobs.com/wday/cxs/{tenant}/{site}/jobs
    Slug format: tenant.server.site (e.g., "salesforce.wd12.External_Career_Site")
    Pagination via offset/limit in POST body. Per-job detail fetches for full descriptions.
    """

    def __init__(self):
        super().__init__("workday", WORKDAY_COMPANIES)

    def _parse_slug(self, slug: str):
        """Parse tenant.server.site slug into components."""
        parts = slug.split(".")
        if len(parts) != 3:
            raise ValueError(f"Invalid Workday slug format: {slug} (expected tenant.server.site)")
        return parts[0], parts[1], parts[2]

    def _base_url(self, tenant: str, server: str) -> str:
        return f"https://{tenant}.{server}.myworkdayjobs.com"

    async def fetch_company_jobs(self, company_name: str, slug: str) -> List[ScrapedJob]:
        tenant, server, site = self._parse_slug(slug)
        base_url = self._base_url(tenant, server)
        jobs_url = f"{base_url}/wday/cxs/{tenant}/{site}/jobs"

        headers = {
            "Accept": "application/json",
            "Content-Type": "application/json",
        }

        jobs: List[ScrapedJob] = []
        offset = 0
        max_per_company = 500  # Cap per company to avoid overwhelming large employers

        async with httpx.AsyncClient() as client:
            while len(jobs) < max_per_company:
                payload = {
                    "appliedFacets": {},
                    "limit": PAGE_SIZE,
                    "offset": offset,
                    "searchText": "",
                }

                response = await client.post(jobs_url, json=payload, headers=headers, timeout=30.0)
                response.raise_for_status()
                data = response.json()

                job_postings = data.get("jobPostings", [])
                if not job_postings:
                    break

                total = data.get("total", 0)

                # Fetch details for each job (Workday listing endpoint has minimal info)
                for posting in job_postings:
                    if len(jobs) >= max_per_company:
                        break

                    job = await self._fetch_and_parse_job(
                        client, base_url, tenant, site, posting, company_name, headers
                    )
                    if job:
                        jobs.append(job)

                    await self.rate_limit()

                offset += PAGE_SIZE
                if offset >= total:
                    break

                await self.rate_limit()

        return jobs

    async def _fetch_and_parse_job(
        self,
        client: httpx.AsyncClient,
        base_url: str,
        tenant: str,
        site: str,
        posting: dict,
        company_name: str,
        headers: dict,
    ) -> Optional[ScrapedJob]:
        """Fetch job details and parse into ScrapedJob."""
        external_path = posting.get("externalPath", "")
        title = posting.get("title", "").strip()
        if not external_path or not title:
            return None

        # Build a stable ID from the path
        job_id = external_path.strip("/").split("/")[-1] if external_path else None
        if not job_id:
            return None

        # Fetch full job details
        detail_url = f"{base_url}/wday/cxs/{tenant}/{site}{external_path}"
        description = ""
        location = ""
        posted_at = None
        job_type_str = ""

        try:
            detail_resp = await client.get(detail_url, headers=headers, timeout=15.0)
            if detail_resp.status_code == 200:
                detail = detail_resp.json()
                job_detail = detail.get("jobPostingInfo", {})

                # Description
                description = self.strip_html(job_detail.get("jobDescription", ""))

                # Additional info
                additional = job_detail.get("additionalInformation", "")
                if additional:
                    description += "\n\n" + self.strip_html(additional)

                # Location
                location = job_detail.get("location", "")

                # Job type
                job_type_str = job_detail.get("timeType", "")

                # Posted date
                posted_str = job_detail.get("postedOn") or job_detail.get("startDate")
                if posted_str:
                    try:
                        posted_at = datetime.fromisoformat(posted_str.replace("Z", "+00:00"))
                    except (ValueError, TypeError):
                        pass
        except Exception as e:
            logger.debug(f"[workday] Failed to fetch details for {job_id}: {e}")
            # Use listing data as fallback
            location = posting.get("locationsText", "")

        # Fallback location from listing
        if not location:
            location = posting.get("locationsText", "")

        # Remote detection
        remote = self.detect_remote(f"{description} {title}", location)

        # Job type
        job_type = self.normalize_job_type(job_type_str)

        # Apply URL
        apply_url = posting.get("externalPath", "")
        if apply_url:
            apply_url = f"{base_url}{apply_url}"

        return self.make_job(
            external_id=str(job_id),
            title=title,
            company_name=company_name,
            description=description,
            apply_url=apply_url,
            location=location,
            remote=remote,
            job_type=job_type,
            posted_at=posted_at,
        )
